src/internal/i18n/provider.tsx

```diff
@@ -0,0 +1,147 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useCallback, useContext, useMemo } from 'react';
+import IntlMessageFormat from 'intl-messageformat';
+import { MessageFormatElement } from '@formatjs/icu-messageformat-parser';
+
+import { InternalI18nContext, FormatFunction, CustomHandler } from './context';
+
+export interface I18nProviderProps {
+  messages: ReadonlyArray<I18nProviderProps.Messages>;
+  locale?: string;
+  children: React.ReactNode;
+}
+
+export namespace I18nProviderProps {
+  export interface Messages {
+    [namespace: string]: {
+      [locale: string]: {
+        [component: string]: {
+          [key: string]: string | MessageFormatElement[];
+        };
+      };
+    };
+  }
+}
+
+/**
+ * Context to send parent messages down to child I18nProviders. This isn't
+ * included in the InternalI18nContext to avoid components from depending on
+ * MessageFormatElement types.
+ */
+const I18nMessagesContext = React.createContext<I18nProviderProps.Messages>({});
+
+export function I18nProvider({ messages: messagesArray, locale: providedLocale, children }: I18nProviderProps) {
+  // Keep a cache per render for improved performance. Generally, it's expected
+  // to have an I18nProvider pretty high up the render tree and for it to
+  // rarely rerender, so it should stick around and offer performance gains.
+  const cache = useMemo(() => new Map<string, unknown>(), []);
+  cache.clear();
+
+  // The provider accepts an array of configs. We merge parent messages and
+  // flatten the tree early on so that accesses by key are simpler and faster.
+  const parentMessages = useContext(I18nMessagesContext);
+  const messages = useMemo(() => mergeMessages([parentMessages, ...messagesArray]), [parentMessages, messagesArray]);
+
+  // If a locale isn't provided, we can try and guess from the html lang,
+  // and lastly default to English. Locales have a recommended case, but are
+  // matched case-insensitively, so we lowercase it internally.
+  const locale = (providedLocale || document?.documentElement.lang || 'en').toLowerCase();
```

Comment:
Should it reuse the stuff from this file? https://github.com/cloudscape-design/components/blob/main/src/internal/utils/locale/normalize-locale.ts

Reply:
How about risking date picker using one locale for the months and another for all provided strings?

Reply:
I'm in favor of aligning the i18n provider with whatever Date picker uses. But I'm fine with shipping this as-is for the Beta. Please make sure that we revisit this for the "proper" launch, though.